File: js/seed.js
// Demo seed against Supabase. Inserts vacancies, candidates, interviews,
// and lets DB triggers populate history + activity_logs + employees.

import { sb } from './lib/supabase.js';
import { unwrap } from './lib/errors.js';
import { createVacancy } from './services/vacancies.js';
import { createCandidate, moveCandidate } from './services/candidates.js';
import { scheduleInterview } from './services/interviews.js';

export async function seedDemoData() {
  // Skip if any vacancies already exist.
  const { count } = await sb.from('vacancies').select('*', { count: 'exact', head: true });
  if (count && count > 0) return false;

  const vacancies = [];
  for (const v of [
    { title: 'Senior Backend Engineer', department: 'Engineering', location: 'Bengaluru, IN', hiringManager: 'Priya Raman', openings: 2, priority: 'high',   targetClose: daysFromNow(14, true), skills: ['Python','PostgreSQL','AWS','Kubernetes'] },
    { title: 'Product Designer',        department: 'Design',      location: 'Remote',        hiringManager: 'Alex Hughes', openings: 1, priority: 'medium', targetClose: daysFromNow(21, true), skills: ['Figma','UX','Prototyping'] },
    { title: 'Data Scientist',          department: 'Data',        location: 'London, UK',    hiringManager: 'Rahul Nair',  openings: 1, priority: 'high',   targetClose: daysFromNow(10, true), skills: ['Python','PyTorch','SQL'] },
    { title: 'Technical Recruiter',     department: 'People',      location: 'New York, US',  hiringManager: 'Sara Owens',  openings: 1, priority: 'low',    targetClose: daysFromNow(30, true), skills: ['Sourcing','ATS','Greenhouse'] },
  ]) {
    vacancies.push(await createVacancy(v));
  }

  const names = [
    ['Ananya Iyer','ananya.iyer@example.com', ['Python','Django','PostgreSQL','AWS']],
    ['Marcus Hale','marcus.hale@example.com', ['Go','Kubernetes','Terraform','AWS']],
    ['Yuki Tanaka','yuki.tanaka@example.com', ['Python','Airflow','Snowflake','SQL']],
    ['Isla Fernández','isla.fernandez@example.com', ['Figma','UX','A/B Testing']],
    ['Kenji Park','kenji.park@example.com', ['React','TypeScript','Node.js']],
    ['Zoe Lewis','zoe.lewis@example.com', ['PyTorch','TensorFlow','Python']],
    ['Rahul Mehta','rahul.mehta@example.com', ['Sourcing','Greenhouse','ATS']],
    ['Nadia Khan','nadia.khan@example.com', ['Java','Spring','Microservices']],
    ['Oliver Schmidt','oliver.schmidt@example.com', ['Python','FastAPI','Redis']],
    ['Chen Wei','chen.wei@example.com', ['Figma','Product Management','UI']],
    ['Farah Aziz','farah.aziz@example.com', ['SQL','BigQuery','Python']],
    ['Leo Romano','leo.romano@example.com', ['React','Next.js','TypeScript']],
  ];
  const stages = ['sourced','sourced','screening','screening','interview','interview','offer','rejected','hired','screening','interview','sourced'];

  const created = [];
  for (let i = 0; i < names.length; i++) {
    const [name, email, skills] = names[i];
    const vac = vacancies[i % vacancies.length];
    const c = await createCandidate({
      name, email,
      phone: randomPhone(),
      location: vac.location,
      headline: headlineFor(skills),
      skills,
      experienceYears: 2 + (i % 8),
      vacancyId: vac.id,
      source: i % 2 ? 'LinkedIn' : 'Referral',
      rating: (i % 5) + 1,
    });
    created.push(c);
    // Walk the candidate through stages so triggers populate history + activity.
    if (stages[i] !== 'sourced') await moveCandidate(c.id, stages[i]);
  }

  // A few interviews
  for (let i = 0; i < 5; i++) {
    await scheduleInterview({
      candidateId: created[i].id,
      interviewer: ['Priya Raman','Alex Hughes','Rahul Nair','Sara Owens'][i % 4],
      type: ['screening','technical','culture','final'][i % 4],
      scheduledAt: hoursFromNow((i + 1) * 20),
      status: 'scheduled',
    });
  }

  return true;
}

export async function resetAllData() {
  // Order matters: respect FK direction.
  for (const t of ['payroll','attendance','employees','interviews','candidates','vacancies','activity_logs']) {
    await unwrap(sb.from(t).delete().neq('id', '00000000-0000-0000-0000-000000000000'), `Failed clearing ${t}`);
  }
}

function daysFromNow(d, dateOnly = false) {
  const t = new Date(Date.now() + d * 86400000);
  return dateOnly ? t.toISOString().slice(0, 10) : t.toISOString();
}
function hoursFromNow(hh) { return new Date(Date.now() + hh * 3600000).toISOString(); }
function randomPhone() { return '+1 ' + Math.floor(200 + Math.random() * 800) + '-' + Math.floor(1000 + Math.random() * 9000); }
function headlineFor(skills) {
  if (skills.some(s => /figma|ux|ui|product/i.test(s)))    return 'Senior Product Designer';
  if (skills.some(s => /torch|tensor|scikit/i.test(s)))    return 'Machine Learning Engineer';
  if (skills.some(s => /sourcing|ats|greenhouse/i.test(s))) return 'Talent Acquisition Partner';
  if (skills.some(s => /react|next|typescript/i.test(s))) return 'Full-stack Engineer';
  return 'Senior Backend Engineer';
}

File: server/seed.py
"""Populate the DB with realistic demo data so the dashboard feels alive."""
from __future__ import annotations

import json
import random
import sqlite3
from datetime import datetime, timedelta, timezone

from . import db


def _iso(dt: datetime) -> str:
    return dt.astimezone(timezone.utc).isoformat(timespec="seconds")


def seed_demo_data() -> bool:
    with db.connect() as c:
        if c.execute("SELECT COUNT(*) FROM vacancies").fetchone()[0] > 0:
            return False
        if c.execute("SELECT COUNT(*) FROM candidates").fetchone()[0] > 0:
            return False

    now = datetime.now(timezone.utc)

    vacancy_defs = [
        ("Senior Backend Engineer", "Engineering", "Bengaluru, IN", "Priya Raman", 2, "High",   14, ["Python","PostgreSQL","AWS","Kubernetes"]),
        ("Product Designer",         "Design",      "Remote",       "Alex Hughes", 1, "Medium", 21, ["Figma","UX","Prototyping"]),
        ("Data Scientist",           "Data",        "London, UK",   "Rahul Nair",  1, "High",   10, ["Python","PyTorch","SQL"]),
        ("Technical Recruiter",      "People",      "New York, US", "Sara Owens",  1, "Low",    30, ["Sourcing","ATS","Greenhouse"]),
    ]

    vacancies = []
    for title, dept, loc, mgr, openings, prio, days_out, skills in vacancy_defs:
        v = db.create_vacancy({
            "title": title,
            "department": dept,
            "location": loc,
            "hiring_manager": mgr,
            "openings": openings,
            "priority": prio,
            "status": "Open",
            "target_close": _iso(now + timedelta(days=days_out)),
            "skills": skills,
        })
        vacancies.append(v)

    candidate_defs = [
        ("Ananya Iyer",     "ananya.iyer@example.com",     ["Python","Django","PostgreSQL","AWS"]),
        ("Marcus Hale",     "marcus.hale@example.com",     ["Go","Kubernetes","Terraform","AWS"]),
        ("Yuki Tanaka",     "yuki.tanaka@example.com",     ["Python","Airflow","Snowflake","SQL"]),
        ("Isla Fernandez",  "isla.fernandez@example.com",  ["Figma","UX","A/B Testing"]),
        ("Kenji Park",      "kenji.park@example.com",      ["React","TypeScript","Node.js"]),
        ("Zoe Lewis",       "zoe.lewis@example.com",       ["PyTorch","TensorFlow","Python"]),
        ("Rahul Mehta",     "rahul.mehta@example.com",     ["Sourcing","Greenhouse","ATS"]),
        ("Nadia Khan",      "nadia.khan@example.com",      ["Java","Spring","Microservices"]),
        ("Oliver Schmidt",  "oliver.schmidt@example.com",  ["Python","FastAPI","Redis"]),
        ("Chen Wei",        "chen.wei@example.com",        ["Figma","Product Management","UI"]),
        ("Farah Aziz",      "farah.aziz@example.com",      ["SQL","BigQuery","Python"]),
        ("Leo Romano",      "leo.romano@example.com",      ["React","Next.js","TypeScript"]),
    ]
    stages = ["sourced","sourced","screening","screening","interview","interview",
              "offer","rejected","hired","screening","interview","sourced"]

    flow = ["sourced","screening","interview","offer","hired"]
    rng = random.Random(42)

    for i, (name, email, skills) in enumerate(candidate_defs):
        vac = vacancies[i % len(vacancies)]
        stage = stages[i]
        cand = db.create_candidate({
            "name": name,
            "email": email,
            "phone": f"+1 {rng.randint(200, 999)}-{rng.randint(1000, 9999)}",
            "location": vac["location"],
            "headline": _headline_for(skills),
            "skills": skills,
            "experience_years": 2 + (i % 8),
            "vacancy_id": vac["id"],
            "source": "LinkedIn" if i % 2 else "Referral",
            "rating": (i % 5) + 1,
            "stage": "sourced",
        })
        _backdate(cand["id"], stage, now, flow, rng)
        _force_stage(cand["id"], stage)

    cands_for_interviews = db.list_candidates()[:5]
    for i, c in enumerate(cands_for_interviews):
        db.create_interview({
            "candidate_id": c["id"],
            "interviewer": ["Priya Raman", "Alex Hughes", "Rahul Nair", "Sara Owens"][i % 4],
            "type": ["Screening", "Technical", "Culture", "Final"][i % 4],
            "date": _iso(now + timedelta(hours=(i + 1) * 20)),
            "status": "Scheduled",
        })

    return True


def _headline_for(skills: list[str]) -> str:
    joined = " ".join(skills).lower()
    if any(t in joined for t in ("figma", "ux", "ui", "product")):
        return "Senior Product Designer"
    if any(t in joined for t in ("torch", "tensor", "scikit")):
        return "Machine Learning Engineer"
    if any(t in joined for t in ("sourcing", "ats", "greenhouse")):
        return "Talent Acquisition Partner"
    if any(t in joined for t in ("react", "next", "typescript")):
        return "Full-stack Engineer"
    return "Senior Backend Engineer"


def _backdate(candidate_id: str, stage: str, now: datetime, flow: list[str], rng: random.Random) -> None:
    """Rewrite candidate history to create realistic time-in-stage data."""
    with db.connect() as c:
        c.execute("DELETE FROM candidate_history WHERE candidate_id = ?", (candidate_id,))
        if stage in flow:
            idx = flow.index(stage)
            for i in range(idx + 1):
                offset = (idx - i + 1) * (3 + rng.random() * 5)
                at = _iso(now - timedelta(days=offset))
                frm = flow[i - 1] if i else None
                c.execute(
                    "INSERT INTO candidate_history (candidate_id, stage, from_stage, at) VALUES (?, ?, ?, ?)",
                    (candidate_id, flow[i], frm, at),
                )
        elif stage == "rejected":
            c.execute("INSERT INTO candidate_history (candidate_id, stage, from_stage, at) VALUES (?, ?, ?, ?)",
                      (candidate_id, "sourced", None, _iso(now - timedelta(days=12))))
            c.execute("INSERT INTO candidate_history (candidate_id, stage, from_stage, at) VALUES (?, ?, ?, ?)",
                      (candidate_id, "screening", "sourced", _iso(now - timedelta(days=9))))
            c.execute("INSERT INTO candidate_history (candidate_id, stage, from_stage, at) VALUES (?, ?, ?, ?)",
                      (candidate_id, "rejected", "screening", _iso(now - timedelta(days=4))))


def _force_stage(candidate_id: str, stage: str) -> None:
    with db.connect() as c:
        c.execute("UPDATE candidates SET stage = ? WHERE id = ?", (stage, candidate_id))
